### PR TITLE
feat: add support for SEP-10 v3.1.0.

### LINKED ISSUE #315: SEP-10 v3.1 Changes (from the SDF)
### TL;DR

Implement the changes described the [SEP-10 3.1 commit](https://github.com/stellar/stellar-protocol/commit/6c8c9cf6685c85509835188a136ffb8cd6b9c11c).

This likely involves adding an optional argument to the SEP-10 utility functions for adding an additional Manage Data operation containing the domain of the SEP-10 server (`web_auth_domain`).

### Background & Description

Thank you for implementing [SEP-10 v3.0](9d121f9), which reintroduced verification of the home domain of the service requiring SEP-10 authentication, ensuring that signed challenges cannot be used for any other service.

SEP-10 v3.1 adds an optional Manage Data operation that includes the home domain of the service issuing SEP-10 challenges. This allows clients and servers to verify the issuer of the challenge.

### PR DESCRIPTION
**Pull request by overcat**
resolve #315 

- Updates the SEP-10 utility function parameters to support [SEP-10 v3.1](https://github.com/stellar/stellar-protocol/commit/6c8c9cf6685c85509835188a136ffb8cd6b9c11c)
  - A new required `webAuthDomain` parameter was added to the following functions
    - `Sep10Challenge#newChallenge(KeyPair, Network, String, String, String, TimeBounds)`
    - `Sep10Challenge#readChallengeTransaction(String, String, Network, String, String)`
    - `Sep10Challenge#readChallengeTransaction(String, String, Network, String[], String)`
    - `Sep10Challenge#verifyChallengeTransactionSigners(String, String, Network, String, String, Set)`
    - `Sep10Challenge#verifyChallengeTransactionSigners(String, String, Network, String[], String, Set)`
    - `Sep10Challenge#verifyChallengeTransactionThreshold(String, String, Network, String[], String, int, Set)`
    - `Sep10Challenge#verifyChallengeTransactionThreshold(String, String, Network, String, String, int, Set)`
  - The `webAuthDomain` parameter is expected to match the value of the Manage Data operation with the 'web_auth_domain' key, if present.

### CODE REVIEW

**Review by tamirms — APPROVED** [review]
this looks good to me but I'll yield to @leighmcculloch for the final approval

File: src/main/java/org/stellar/sdk/Sep10Challenge.java [hunk 1]
```diff
@@ -63,20 +70,21 @@ public static Transaction newChallenge(
...
    * @param network         The network to connect to for verifying and retrieving.
    * @param domainNames     An array of home domains, one of which is expected to be included in the first Manage Data operation's string key.
+   * @param webAuthDomain   The home domain that is expected to be included as the value of the Manage Data operation with the 'web_auth_domain' key. If no such operation is included, this parameter is not used.
    * @return {@link ChallengeTransaction}, the decoded transaction envelope and client account ID contained within.
    * @throws InvalidSep10ChallengeException If the SEP-0010 validation fails, the exception will be thrown.
    * @throws IOException                    If read XDR string fails, the exception will be thrown.
    */
-  public static ChallengeTransaction readChallengeTransaction(String challengeXdr, String serverAccountId, Network network, String[] domainNames) throws InvalidSep10ChallengeException, IOException {
+  public static ChallengeTransaction readChallengeTransaction(String challengeXdr, String serverAccountId, Network network, String[] domainNames, String webAuthDomain) throws InvalidSep10ChallengeException, IOException {
```
Comment (leighmcculloch) [inline src/main/java/org/stellar/sdk/Sep10Challenge.java:87]: In the Go SDK we need to put the webauth domain before the home domains to preserve an odd but awkward case that can break applications without them realizing, so I think there would be value in having all SDKs have that field first for consistency. But it also doesn't really matter. cc @JakeUrban

**Review by leighmcculloch — APPROVED** [review]
Looks good to me, although one concern. @JakeUrban could you chime in?

File: src/main/java/org/stellar/sdk/Sep10Challenge.java (hunk 1 above)
Comment (JakeUrban) [reply src/main/java/org/stellar/sdk/Sep10Challenge.java:87]: I don't think it makes sense to try to keep function signatures consistent across languages. I'm pretty sure that ship has already sailed anyway. The change made looks good to me 👍

**Review by JakeUrban — APPROVED** [review]
Looks good to me, thanks @overcat

File: src/main/java/org/stellar/sdk/Sep10Challenge.java (hunk 1 above)
Comment (leighmcculloch) [reply src/main/java/org/stellar/sdk/Sep10Challenge.java:87]: 👍🏻

**Comment by tamirms** [comment]
@overcat could you try rebasing your branch on top of master? we need to retrigger circle ci so that the PR can be merged

**Comment by overcat** [comment]
Thank you all.

**Pull request merged**